pkg/sources/forward/data_forward.go

```diff
@@ -258,49 +259,57 @@ func (df *DataForward) forwardAChunk(ctx context.Context) error {
 		return nil
 	}
 
-	// reset the idle handler because we have read messages
-	df.srcIdleHandler.Reset()
-	metrics.ReadDataMessagesCount.With(map[string]string{
-		metrics.LabelVertex:             df.vertexName,
-		metrics.LabelPipeline:           df.pipelineName,
-		metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
-		metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
-		metrics.LabelPartitionName:      df.reader.GetName()},
-	).Add(float64(len(readMessages)))
-
-	metrics.ReadMessagesCount.With(map[string]string{
+	// Only expose read processing time metric when there are messages and no error
+	metrics.ReadProcessingTime.With(map[string]string{
 		metrics.LabelVertex:             df.vertexName,
 		metrics.LabelPipeline:           df.pipelineName,
 		metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
 		metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
 		metrics.LabelPartitionName:      df.reader.GetName(),
-	}).Add(float64(len(readMessages)))
+	}).Observe(float64(time.Since(readStart).Microseconds()))
+	// reset the idle handler because we have read messages
+	df.srcIdleHandler.Reset()
 
 	// store the offsets of the messages we read from source
-
-	// store the offsets of the messages we read from ISB
 	var readOffsets = make([]isb.Offset, len(readMessages))
 	for idx, m := range readMessages {
-		totalBytes += len(m.Payload)
+		totalBytes := len(m.Payload)
+
```

Comment:
ItoA replica out of the loop, and do the partitionIdx here to improve performance?